File: algorithm/64. Minimum Path Sum/64.py
class Solution(object):
    def minPathSum(self, grid):
        """
        :type grid: List[List[int]]
        :rtype: int
        """
        m=len(grid)
        n=len(grid[0])
        path = [[grid[0][0] for i in range(n)] for x in range(m)]
        for i in range(1,m):
            path[i][0]= path[i-1][0] + grid[i][0]
        for i in range(1,n):
            path[0][i]= path[0][i-1]+ grid[0][i]
        for i in range(1,m):
            for j in range(1,n):
                path[i][j] = min(path[i-1][j], path[i][j-1])+grid[i][j]
        return path[m-1][n-1]

File: algorithm/64. Minimum Path Sum/64.cpp
class Solution {
public:
    int minPathSum(vector<vector<int>>& grid) {
        int m = grid.size();
        int n = grid[0].size();
        vector<vector<int> > path(m, vector<int>(n, grid[0][0]));
        for(int i=1; i< n; i++)
        {
            path[0][i]=grid[0][i]+path[0][i-1];
        }
        for(int i=1; i< m; i++)
        {
            path[i][0]=grid[i][0]+path[i-1][0];
        }
        for (int i=1; i<m; i++)
        {
            for(int j=1; j<n; j++)
            {
                if(path[i-1][j]<path[i][j-1])
                {
                    path[i][j]= grid[i][j]+ path[i-1][j];
                }
                else{
                    path[i][j] = grid[i][j] + path[i][j-1];
                }
            }
        }
        return path[m-1][n-1];
    }
};
